pylint/checkers/variables.py

```diff
@@ -665,6 +673,146 @@ def get_next_to_consume(self, node: nodes.Name) -> list[nodes.NodeNG] | None:
+        tests that evaluate false.
+
+        Don't identify a node if there is a collectively exhaustive set of paths
+        that define the name, raise, or return (e.g. every if/else branch).
+        """
+        uncertain_nodes = []
+        for other_node in found_nodes:
+            if in_type_checking_block(other_node):
+                continue
+
+            if not isinstance(other_node, nodes.AssignName):
+                continue
+
+            closest_if = utils.get_node_first_ancestor_of_type(other_node, nodes.If)
+            if closest_if is None:
+                continue
+            if node.frame() is not closest_if.frame():
+                continue
+            if closest_if is not None and closest_if.parent_of(node):
+                continue
+            if any(
+                if_node.parent_of(closest_if)
+                for if_node in self._if_nodes_deemed_uncertain
+            ):
+                uncertain_nodes.append(other_node)
+                continue
+
+            # All inferred values must test false
+            if isinstance(closest_if.test, nodes.NamedExpr):
+                test = closest_if.test.value
+            else:
+                test = closest_if.test
+            all_inferred = utils.infer_all(test)
+            if not all_inferred or not all(
+                isinstance(inferred, nodes.Const) and not inferred.value
+                for inferred in all_inferred
+            ):
+                continue
+
+            # Name defined in every if/else branch
+            if NamesConsumer._exhaustively_define_name_raise_or_return(
+                other_node.name, closest_if
+            ):
+                continue
+
+            # Name defined somewhere else during a loop
+            if isinstance(closest_if.test, nodes.Name):
+                loop_names = set()
+                other_node_frame = other_node.frame(future=True)
+                for_ancestor = None
+                for ancestor in other_node.node_ancestors():
+                    if ancestor.frame() is not other_node_frame:
+                        break
+                    if isinstance(ancestor, nodes.For):
+                        loop_names.update(
+                            {
+                                a.name
+                                for a in ancestor.target.nodes_of_class(
+                                    nodes.AssignName
+                                )
+                            }
+                        )
+                        for_ancestor = ancestor
+                if closest_if.test.name in loop_names:
+                    continue
+                if for_ancestor is not None and isinstance(
+                    other_node, nodes.AssignName
+                ):
+                    if any(
+                        an.name == other_node.name
+                        for an in for_ancestor.nodes_of_class(nodes.AssignName)
+                    ):
+                        continue
```

Comment:
I feel like creating a function here would make sense, but the ``continue`` design makes it a little hard. To be fair I find the ``_uncertain_nodes_in_false_tests`` function a little complicated, Wild proposition but maybe we could loop on a list of conditional instead of having a conditional then a continue ?

Reply:
I know it's long! Thanks for reviewing. I think the issue with a list of conditionals is that without a walrus operator I'd be repeating lookups like `closest_if` and `other_node_frame` unnecessarily. Happy to discuss other approaches for factoring, though.

Reply:
Let's revisit when python 3.7 has reached end of life then 😄